src/engine/NetworkLevelReasoner.cpp

```diff
@@ -330,6 +417,108 @@ void NetworkLevelReasoner::updateVariableIndices( const Map<unsigned, unsigned>
     }
 }
 
+void NetworkLevelReasoner::obtainInputBounds()
+{
+    ASSERT( _tableau );
+
+    for ( unsigned i = 0; i < _layerSizes[0]; ++i )
+    {
+        unsigned varIndex = _indexToActivationResultVariable[Index( 0, i )];
+        _lowerBoundsActivations[0][i] = _tableau->getLowerBound( varIndex );
+        _upperBoundsActivations[0][i] = _tableau->getUpperBound( varIndex );
+    }
+}
+
+void NetworkLevelReasoner::setTableau( const ITableau *tableau )
+{
+    _tableau = tableau;
+}
+
+void NetworkLevelReasoner::intervalArithmeticBoundPropagation()
+{
+    for ( unsigned i = 1; i < _numberOfLayers; ++i )
+    {
+        for ( unsigned j = 0; j < _layerSizes[i]; ++j )
+        {
+            Index index( i, j );
+            double lb = _bias.exists( index ) ? _bias[index] : 0;
+            double ub = _bias.exists( index ) ? _bias[index] : 0;
+
+            for ( unsigned k = 0; k < _layerSizes[i - 1]; ++k )
+            {
+                double previousLb = _lowerBoundsActivations[i - 1][k];
+                double previousUb = _upperBoundsActivations[i - 1][k];
+                double weight = _weights[i - 1][k * _layerSizes[i] + j];
+
+                if ( weight > 0 )
+                {
+                    lb += weight * previousLb;
+                    ub += weight * previousUb;
+                }
+                else
+                {
+                    lb += weight * previousUb;
+                    ub += weight * previousLb;
+                }
+            }
+
+            _lowerBoundsWeightedSums[i][j] = lb;
+            _upperBoundsWeightedSums[i][j] = ub;
+
+            if ( i != _numberOfLayers - 1 )
+            {
+                // Apply activation function
+                ASSERT( _neuronToActivationFunction.exists( index ) );
+
+                switch ( _neuronToActivationFunction[index] )
+                {
+                case ReLU:
+                    _lowerBoundsActivations[i][j] = lb > 0 ? lb : 0;
+                    _upperBoundsActivations[i][j] = ub;
+                    break;
+
+                default:
+                    ASSERT( false );
```

Comment:
is NLR going to be used only for networks with Relus?

Reply:
OK. So, how about if we don't have the assertion and then the code won't do anything on other networks. Right?

Reply:
It might produce erroneous results, if it doesn't take into account the effects of other functions. For example, if we have absolute value and we let a negative value through.

Reply:
i see. hmm...

I am worried that in the production mode, we will not detect this. Perhaps, we should throw an exception here.

Reply:
Good point. Fixed.